# IN1160 – Oblig 1a: Vektorrom

**Våren 2026**

Det er en god idé å lese gjennom hele oppgavesettet før dere setter i gang.
Dersom dere har spørsmål så kan dere:

- spørre i gruppetimene,
- spørre på Discourse
- eller sende epost til in1160-support@ifi.uio.no dersom alternativene over av en eller annen grunn ikke passer for spørsmålet.


## Innlevering

Oppgaven leveres innen 11.02. klokken 23:59 i [Devilry](https://devilry.ifi.uio.no/). 

Innleveringen skal bestå av én Jupyter notebook med både kode og tilhørende forklaringer. **Før innlevering skal du kjøre gjennom hele notebooken, før du lagrer siste gang. Den bør kjøre uten å feile og vise den grafikken og de utskriftene som skal være med.**
Skulle dere av tekniske grunner ikke kunne besvare alle spørsmålene i en notebook kan dere i denne obligen også levere vanlige Python-filer, sammen med en PDF som inneholder grafikker og resultater fra kjøring, samt deres tekst-besvarelser. Ta kontakt med emnestaben, så vi får ordnet eventuelle problemer før oblig 1b.

Vi understreker at innlevering av kode alene ikke er nok for å bestå oppgaven – vi forventer at notebooken også skal inneholde kommentarer (på norsk eller engelsk) på hva dere har gjort og begrunnelser for valgene dere har tatt underveis.
La enhver oblig bli en trening i å formidle forskning. Bruk helst hele setninger, og matematiske formler om nødvendig. Resultater skal presenteres i tabeller på en oversiktlig måte.
Det å forklare med egne ord, bruke begreper vi har gått gjennom på forelesningene og å forklare og reflektere over løsningene deres er en viktig del av læringsprosessen – ta det på alvor!

Når det gjelder bruk av generative prateroboter (ChatGPT og lignende): Dere kan bruke dem som en "sparringspartner", for eksempel for å forklare noe dere ikke helt har forstått. Dere har imidlertid ikke lov til å bruke dem til å generere løsninger (enten delvis eller fullstendig) til noen av oppgavene. Funksjoner for automatisk skriving av kode, som Copilot i VS Code, må derfor også være deaktivert mens dere jobber på obligen.  
Bruker dere KI-verktøy vil vi også at dere kort beskriver hvordan dere har brukt dem under arbeidet med oppgaven.

Det er ikke mulighet for omlevering av obliger som ikke bestås.


## Oppsett

Denne første obligen skal blant annet hjelpe dere med å komme i gang med verktøyene vi kommer til å bruke gjennom semesteret. Sørg for å aktivere IN1160-miljøet, så bør dere kunne kjøre kodeblokken under. Får dere noen feilmeldinger, oppfordres dere til å møte opp i lab- og gruppetimene for teknisk hjelp.


In [ ]:
from helpers_1a import sanity_check

sanity_check()

## Bakgrunn

I denne innleveringen skal vi jobbe med datasettet _Norsk aviskorpus_ (NAK) fra Nasjonalbiblioteket. Dette korpuset består av tekster hentet fra en rekke norske aviser fra perioden 1998-2019.
Det fulle datasettet kan finnes [her](https://www.nb.no/sprakbanken/ressurskatalog/oai-nb-no-sbr-4/).

NAK består av over 1 500 000 ord (~2,6 GB), men vi skal jobbe med en forenklet versjon av datasettet, gitt i fila `NAK_oblig_1a.txt`.

I denne obligen skal dere jobbe med å lage enkle vektorrepresentasjoner av ord. Dere skal gjøre dette basert på hvilke andre ord de forekommer sammen med i setninger i korpuset. Vi skal så se på hvordan vi kan bruke disse vektorene til å finne likheter mellom ord ved bruk av cosinuslikhet.

I oppgavene som følger skal dere:

1. gjøre preprosessering av tekstene
2. representere ord som vektorer i en vektorrom-modell
3. undersøke likhet mellom ord gjennom vektorene


## Oppgave 1 – Forberede data


### Oppgave 1.1 – Laste inn datasett (1 poeng)

I denne oppgaven skal dere laste inn NAK-datasettet fra tekstfila `NAK_oblig_1a.txt`. Dette skal gjøres med funksjonen `load_dataset()`.
Denne tar filnavnet til datasettet (som skal ligge i samme mappe som denne notebook-fila) som argument og skal returnere en liste med setningstrenger.

Skriv ferdig funksjonen `load_dataset()` i cellen under og last inn datasettet:


In [1]:
# Oppgave 1.1
def load_dataset(dataset_path):
    """
    Argumenter:
    - dataset_path : En streng som spesifiserer hvor vi finner datasettet.

    Returnerer:
    - dataset : En liste med setninger.
    """

    # Din kode her
    file = open(dataset_path, 'r', encoding="utf-8")
    dataset = file.readlines()
    return dataset


In [2]:
# Bruk load_dataset() til å laste inn datasettet her:
sentences = load_dataset("NAK_oblig_1a.txt")

### Oppgave 1.2 – Tokenisering og inspeksjon av dataene (2 poeng)

Vi har nå en liste med setninger, men vi ønsker å utforske de individuelle ordene, så vi må dele opp teksten ytterligere. Denne typen oppdeling av tekst kaller vi for *tokenisering*, fordi hver tekstbit vi mater inn i en modell kalles et *token*. Vi er ofte interessert i å vite hvor mange unike tokens vi har i datasettet vårt. Hvert unike token kaller vi for en *ordtype*, eller bare *type*. Alle ordtypene vi har til sammen kaller vi for *vokabularet* vårt.

I følgende setning er det altså 7 tokens og 5 typer (vi teller "En" og "en" som to forskjellige typer):

> En for alle og alle for en

I cellen under skal dere lage funksjonen `tokenize()` som tokeniserer en setning ved å dele den opp med metoden `.split()`.

Videre skal dere:
- Tokenisere hver setning i datasettet med `tokenize()`. 
  - Dette skal altså resultere i en liste av lister med individuelle ord.
- Rapportere hvor mange ord dere får.
- Rapportere hvor mange ordtyper dere får.


In [3]:
# Oppgave 1.2
def tokenize(text):
    """Tar inn en streng med tekst og returnerer en liste med ord."""
    return text.split()

In [4]:
# Tokeniser datasettet her:
#word_list gir en liste av lister av ord, der hver liste er en setning fra datasettet
word_list = []
#total_word_list er alle ordene fra word_list, men satt i en enkelt liste
total_word_list = []
for i, sentence in enumerate(sentences):
    word_list.append([])
    word_list[i] = tokenize(sentence)
    for word in word_list[i]:
        total_word_list.append(word)

In [5]:
# Beregn antall ord og ordtyper her:
print("Antall ord: ", len(total_word_list))

from collections import Counter
#Bruker klassen Counter fra collections siden det er en lett måte å telle ordtyper på
types = Counter(total_word_list)
print("Antall ordtyper: ", len(types))

Antall ord:  28671744
Antall ordtyper:  657933


### Oppgave 1.3 – Fjerning av stoppord (2 poeng)

Hvis alt er gjort riktig så langt bør dere ha fått ~31 000 000 tokens og ~1 000 000 ordtyper. Det betyr altså at hver type i gjennomsnitt forekommer omtrent 31 ganger i datasettet.
Vi vet derimot (fra [Zipf's lov](https://en.wikipedia.org/wiki/Zipf%27s_law)) at noen relativt få ord forekommer veldig mange ganger, mens veldig mange andre ord forekommer veldig sjeldent. 

Blant de mest vanlige ordene finner vi ord som for eksempel "og", "i" og "det". Siden disse ordene forkommer såpass ofte, bidrar de sjeldent med informasjon som hjelper oss med å skille mellom ulike ord. Vi kan derfor hoppe over disse ordene for å redusere kjøretiden på treningen av modellene våre.  
Denne prosessen kaller vi å fjerne *stoppord*. En liste med norske stoppord (inkludert tegnsetting) er gitt i funksjonen `get_norwegian_stopwords()` i `helpers_1a.py`.

I denne oppgaven skal dere:

- Hente lista med stoppord med `get_norwegian_stopwords()`. 
- Fjerne alle stoppordene fra datasettet. 
  - En setning som *"en katt hadde spist en mus"* skal altså reduseres til *"katt spist mus"*.
- Rapportere hvor mange ordtyper og hvor mange tokens det er igjen i datasettet. 
  - Skriv en kort kommentar om hvordan forholdet mellom disse tallene endrer seg.


In [6]:
from helpers_1a import get_norwegian_stopwords

In [7]:
# Oppgave 1.3
stopwords = get_norwegian_stopwords()

for i, sentence in enumerate(word_list):
    word_list[i] = [word for word in word_list[i] if word not in stopwords]

total_word_list = [word for word in total_word_list if word not in stopwords]

#### Oppgave 1.3 – Tekstbesvarelse

_Hvor mange ordtyper og tokens står dere igjen med etter fjerning av stoppordene?_

**Svar:** Fra koden under ender vi med 18357914 ord, og 657768 ordtyper.


In [8]:
print("Antall ord etter fjernte stoppord:", len(total_word_list))

types = Counter(total_word_list)
print("Antall typer etter fjernte stoppord", len(types))

Antall ord etter fjernte stoppord: 18357914
Antall typer etter fjernte stoppord 657768


### Oppgave 1.4 – Fjerning av sjeldne ord (2 poeng)

Som nevnt er det også vanlig at vi har svært mange ord som forekommer svært sjeldent. Blant disse ordene finner vi ofte feilstavinger og ord som ikke har blitt lest riktig fra teksten.
Disse ordene er som oftest heller ikke særlig nyttige for modellene våre og kan derfor også fjernes. 

I denne oppgaven skal dere:

- Lage en liste med de 10 000 ordene som forekommer oftest i datasettet (der stoppord er fjernet).
  - Her kan det være nyttig å benytte seg av klassen `Counter` og metoden `.most_common()`.
- Rapportere hvor mange typer og tokens det er i datasettet etter dette.
  - Hvordan er forholdet mellom disse nå?

Tips: Under utvikling kan det være lurt å begrense vokabularet enda litt mer. Dere kan f.eks. hente de 100 vanligste ordene.


In [9]:
from collections import Counter

In [10]:
# Oppgave 1.4
most_common_10000 = types.most_common(10000)

print("Antall typer av de 10000 mest vanlige ordene: ", len(most_common_10000))

most_common_10000_occurences = 0
for i, word in enumerate(most_common_10000):
    most_common_10000_occurences += most_common_10000[i][1]
print("Antall tokens av de 10000 mest vanlige ordene: ", most_common_10000_occurences)

most_common_10000_words = []
for word in most_common_10000:
    most_common_10000_words.append(word[0])

#Fjerne alle de sjeldne ordene fra datasettet vårt:
for i, sentence in enumerate(word_list):
    word_list[i] = [word for word in sentence if word in most_common_10000_words]

Antall typer av de 10000 mest vanlige ordene:  10000
Antall tokens av de 10000 mest vanlige ordene:  14085239


#### Oppgave 1.4 – Tekstbesvarelse

_Hvor mange typer og tokens står dere igjen med etter fjerning av de sjeldne ordene?_

**Svar:** Fra koden over får vi at vi har 10000 ordtyper, og 14085239 tokens.

## Oppgave 2 – Vektorisering


### Oppgave 2.1 – Fra ord til tall (2 poeng)


Vi har nå fjernet både de vanligste og sjeldneste de ordene fra vokabularet vårt. De resterende ordene burde være en passe mengde for modellen vår.
Modellen vi skal bruke kan derimot ikke bruke strenger som treningsdata, så vi må lage tallrepresentasjoner av dem, der hver type i vokabularet vårt knyttes til et bestemt tall.

For å holde styr på sammenhengen mellom ordene og tall kan vi bruke en ordbok der vi tilordner et tall til hvert ord. For å knytte tallene tilbake til ordene kan vi lage en tilsvarende ordbok med tallet som nøkkel og ordet som verdi.

I denne oppgaven skal dere fylle ut de to ordbøkene `word_to_int` og `int_to_word`. Her kan funksjonen `enumerate()` være nyttig.

In [11]:
# Oppgave 2.1
word_to_int = {}
int_to_word = {}

for i, word in enumerate(most_common_10000):
    word_to_int[word[0]] = i
    int_to_word[i] = word[0]

### Oppgave 2.2 – Bygging av vektorer (5 poeng)

I denne oppgaven skal vi lage vektorrepresentasjoner av ordene vi kom fram til i **1.4**. Vi skal benytte oss av enkle tellevektorer, der vi bruker _konteksten_ et ord befinner seg i til å informere oss om hva slags ord det er snakk om. Intuisjonen for dette er at ord som brukes i samme kontekst typisk har like eller relaterte betydninger:

> "_You shall know a word by the company it keeps_"  
>  – J. R. Firth (1957)

I cellen under finner dere en påbegynt `CountVectorizer`-klasse.
Når det blir opprettet en instans av denne, lager metoden `zeros()` en matrise av nuller som er like lang og like bred som vokabularet vi sender inn. Her representerer hver rad et ord i vokabularet vårt, der f.eks. rad 5 representerer ordet som ble tilordnet tallet 5 i forrige oppgave. Kolonnene fungerer på samme måte. 

Etter å ha opprettet en `CountVectorizer`, skal man kunne kalle på metoden `.fit()`, der man sender inn de tokeniserte tekstdataene fra oppgave **1.2**. Denne sender én og én setning inn til metoden `_add_sentence()`, som dere skal fylle ut.  
Denne metoden skal oppdatere matrisen ved å telle antall ganger to ord forekommer i samme setning. For hvert ord $i$ i setningen, skal vi altså legge til 1 på posisjon $(i, j)$ i matrisen for hvert ord $j$ som også forekommer i setningen.

Når treningen er overstått kan vi hente ut en vektor for et ord ved å bruke metoden `get_vector()`.

I denne oppgaven skal dere:

- Fylle ut metoden `_add_sentence()` slik at matrisen fylles ut på riktig måte.
- Opprette en instans av `CountVectorizer`.
- Trene denne på de tokeniserte dataene.
  - Dette kan ta litt tid avhengig av maskinen dere kjører på.
- Hente ut en vektor for et ord dere selv velger. Finn ut av hvilket annet ord dette forekommer oftest med.


In [12]:
from numpy import zeros

class CountVectorizer:
    def __init__(self, vocab, word_to_int):
        self.vocab = vocab
        self.word_to_int = word_to_int
        self.matrix = self._create_matrix(vocab)

    def _create_matrix(self, vocab):
        """
        Lager en matrise bestående av 0-vektorer. Matrisen er 
        like lang og like bred som vokabularet vårt, slik at
        hver rad representerer et ord, og hver kolonne
        representerer antall ganger et annet ord forekommer
        med rad-ordet.
        """
        matrix = zeros((len(vocab), len(vocab)))
        
        return matrix

    def _add_sentence(self, sentence_tokens):
        """
        Oppdaterer matrisen med observasjoner fra setningen.
        """
        # Fyll ut denne metoden!
        for i, word in enumerate(sentence_tokens):
            for j, word2 in enumerate(sentence_tokens):
                """
                Antar her at det ikke lbir lagt til for posisjon (i, i) for samme ord, at det bar blir lagt
                til hvis samme ord forekommer mer enn en gang i en setning.
                """
                if i != j:
                    self.matrix[self.word_to_int[word]][self.word_to_int[word2]] += 1

    def fit(self, dataset):
        """Trener modellen på `dataset`."""
        for sent in dataset:
            self._add_sentence(sent)

    def get_vector(self, word):
        """Henter vektoren for et gitt ord."""
        word_int = self.word_to_int[word]
        
        return self.matrix[word_int]

In [13]:
# Oppgave 2.2
# Lag en instans av CountVectorizer med våre parametre
vectorizer = CountVectorizer(most_common_10000_words, word_to_int)
# Tren vektorisereren på våre data
vectorizer.fit(word_list)

In [20]:
# Oppgave 2.2
# Hent en vektor for et ord og finn ut av hvilket ord det oftest forekommer sammen med.
import numpy as np
print(int_to_word[np.argmax(vectorizer.get_vector("år"))])

gammel


### Oppgave 2.3 – Vektordimensjonalitet (2 poeng)

Hva kan dere si om vektorene dere nå har laget? Er de høydimensjonale eller ikke? Nevn minst en fordel og en ulempe ved denne typen vektorer.

#### Oppgave 2.3 – Tekstbesvarelse

_Svar her._


### Oppgave 2.4 – Måle distanser med cosinuslikhet (4 poeng)

Vi har nå laget en vektorrommodell av ordene i datasettet vårt, der hver kolonne representerer et ord i vokabularet vårt. Vi skal nå forsøke å tallfeste hvor like to vektorer, så vi kan se om dette stemmer overens med likheter mellom ord.
Dette skal vi gjøre ved bruk av cosinuslikhet (engelsk: _cosine similarity_), der vi måler vinkelen mellom to vektorer.

Cosinuslikheten mellom to vektorer $a$ og $b$ er definert som:

$$
cos(a,\ b) = \frac{a \cdot b}{\|a\|\|b\|}
$$

Her er $a \cdot b$ prikkproduktet av de to vektorene, og $\|a\|$ og $\|b\|$ er normen (lengden) til henholdsvis vektor $a$ og $b$.

Dere skal implementere funksjonen for cosinuslikhet på egenhånd. Det vil si at dere ikke skal importere ferdige funksjoner, men heller implementere dette direkte i Python.
Dere kan derimot bruke pakken `math`, som har en rekke nyttige funksjoner for numeriske beregninger.

I denne oppgaven skal dere:

- Implementere funksjonen `cosine_similarity(a, b)` som tar inn to vektorer og returnerer cosinuslikheten.
- Bruke funksjonen til å finne cosinuslikheten mellom to ord dere selv velger. Skriv en kort kommentar på hva dere observerer.

In [18]:
# Oppgave 2.4
import math

def cosine_similarity(a, b):
    # Regn ut cosinuslikhet her
    dot_product = 0
    a_length_squared = 0
    b_length_squared = 0
    for i, number in enumerate(a):
        dot_product += a[i]*b[i]
        a_length_squared += a[i]**2
        b_length_squared += b[i]**2
    a_length = math.sqrt(a_length_squared)
    b_length = math.sqrt(b_length_squared)
    return dot_product/(a_length*b_length)

In [19]:
# Eksempelbruk:
#Sammenligner på de to vektorene vi brukte i forrige oppgave:
a_vec = vectorizer.get_vector("år")
b_vec = vectorizer.get_vector("gammel")
print(cosine_similarity(a_vec, b_vec))

0.6550051164069103


### Oppgave 2.5 – Finne likheter mellom ord (5 poeng)

Vi skal nå bruke cosinuslikhet til å finne ord som er like hverandre i vektorrommet vårt.
I cellen under finner dere to sett med ord i listene `WORDS` og `WORDS_SUBSET`. For hvert av ordene i `WORDS_SUBSET` skal dere finne de fem ordene fra `WORDS` som er likest, målt med cosinuslikhet.  
Lag en tabell i en markdown-celle der dere setter inn hvert ord i en kolonne og de fem likeste ordene i en annen kolonne. 
Dette kan dere gjøre på følgende måte:

```md
| Ord    | 5 likeste |
| ------ | --------- |
| hund   | ...       |
| salat  | ...       |
| bil    | ...       |
...
```

Diskuter kort resultatene dere får. Virker det som at likere ord får høyere cosinuslikhet? Kom med noen eksempler og/eller moteksempler.

I denne oppgaven skal dere:

- Regne ut cosinuslikhet mellom hvert ord i `WORDS_SUBSET` og alle ordene i `WORDS`.
- Lage en tabell med de fem likeste ordene for hvert ord. 
- Diskutere resultatene deres. 


In [ ]:
WORDS = [
    "gris", "elefant", "hund", "katt", "hest", "bjørn", "kylling", "svin", 
    "fisk", "salat", "brød", "ost", "smør", "melk", "bil", "moped", "sykkel", 
    "båt", "fly", "tog", "sykepleier", "lege", "ingeniør", "sjef", 
    "sjåfør", "designer", "jente", "gutt", "mann", "kvinne", "Oslo", "Trondheim", 
    "Roma", "Egypt", "Mexico", "Galdhøpiggen", "Sørlandet", "Sahara", "Bjørn", 
    "Anne", "Jonas", "Erna", "Donald", "Barack", "Mohamed", "Angela"
]
WORDS_SUBSET = [
    "hund", "salat", "bil", "ingeniør", "sjef", "jente", "Oslo", "Jonas"
]


In [23]:
# Oppgave 2.5
#Lag en ordbok for å holde de fem likeste ordene
word_dictionary = {}
for word in WORDS_SUBSET:
    word_dictionary[word] = {}
    for word2 in WORDS:
        word_vector_1 = vectorizer.get_vector(word)
        word_vector_2 = vectorizer.get_vector(word2)
        similarity = cosine_similarity(word_vector_1, word_vector_2)
        if word != word2:
            """Hvis ordboken ikke har 5 ord for et gitt ord, så blir det
            lagt til ord. Hvis det har 5 ord, blir likheten av et nytt ord sammenlignet med
            likheten til det minst like ordet, og hvis det er større blir det lagt til mens det 
            minste blir fjernet."""
            if len(word_dictionary[word]) < 5:
                word_dictionary[word][word2] = similarity
            else:
                if min(word_dictionary[word].values()) < similarity:
                    word_dictionary[word][word2] = similarity
                    del word_dictionary[word][min(word_dictionary[word], key=word_dictionary[word].get)]

for word in word_dictionary:
    print(word, list(word_dictionary[word].keys()))

hund ['katt', 'bil', 'båt', 'fly', 'Sørlandet']
salat ['kylling', 'fisk', 'brød', 'ost', 'smør']
bil ['hund', 'sykkel', 'båt', 'tog', 'sjåfør']
ingeniør ['hest', 'sykkel', 'lege', 'sjef', 'designer']
sjef ['hest', 'fly', 'lege', 'ingeniør', 'designer']
jente ['bil', 'sjåfør', 'gutt', 'mann', 'kvinne']
Oslo ['fly', 'tog', 'Trondheim', 'Bjørn', 'Anne']
Jonas ['lege', 'sjef', 'Trondheim', 'Bjørn', 'Anne']


#### Oppgave 2.5 – Tekstbesvarelse

_Legg inn tabellen din her. Hva ser du av resultatene?_



**Svar:**
|Ord|5 Likeste|
|---|---------|
|hund|katt, bil, bår, fly, Sørlandet|
|salat|kylling, fisk, brød, ost, smør|
|bil|hund, sykkel, båt, tog, sjåfør|  
|ingeniør|hest, sykkel, lege, sjef, designer|
|sjef|hest, fly, lege, ingeniør, designer|
|jente|bil, sjåfør, gutt, mann, kvinne|
|Oslo|fly, tog, Trondheim, Bjørn, Anne|
|Jonas|lege, sjef, Trondheim, Bjørn, Anne|

Tabellen gir det man ville forventet når det gjelder ord som vil forekomme sammen. For eksempel blir andre transportmiddel nevnt ofte sammen med ordet bil, eller andre yrker nevnes ofte sammen med ord som ingeniør eller sjef.

## NB!

Husk å kjøre gjennom hele notebooken fra toppen av før dere leverer obligen! 
